src/arkode/arkode_forcingstep.c

```diff
@@ -0,0 +1,415 @@
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return ARK_MEM_NULL;
+  }
+  *ark_mem = (ARKodeMem)arkode_mem;
+
+  return forcingStep_AccessStepMem(*ark_mem, __func__, step_mem);
+}
+
+/*---------------------------------------------------------------
+  This routine checks if all required vector operations are
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+static sunbooleantype forcingStep_CheckNVector(const N_Vector y)
+{
+  return y->ops->nvlinearsum != NULL;
+}
+
+/*---------------------------------------------------------------
+  This routine is called just prior to performing internal time
+  steps (after all user "set" routines have been called) from
+  within arkInitialSetup.
+
+  With initialization types FIRST_INIT this routine:
+  - sets the forcing for stepper[1]
+
+  With other initialization types, this routine does nothing.
+  ---------------------------------------------------------------*/
+static int forcingStep_Init(const ARKodeMem ark_mem, const int init_type)
+{
+  ARKodeForcingStepMem step_mem = NULL;
+  int retval = forcingStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return retval; }
+
+  /* assume fixed outer step size */
+  if (!ark_mem->fixedstep)
+  {
+    arkProcessError(ark_mem, ARK_ILL_INPUT, __LINE__, __func__, __FILE__,
+                    "Adaptive outer time stepping is not currently supported");
+    return ARK_ILL_INPUT;
+  }
+
+  /* immediately return if resize or reset */
+  if (init_type == RESIZE_INIT || init_type == RESET_INIT)
+  {
+    return ARK_SUCCESS;
+  }
+
+  ark_mem->interp_degree = 1;
+
+  return ARK_SUCCESS;
+}
+
+/*------------------------------------------------------------------------------
+  This is just a wrapper to call the user-supplied RHS function,
+  f^1(t,y) + f^2(t,y).
+
+  This will be called in one of three 'modes':
+
+     ARK_FULLRHS_START -> called at the beginning of a simulation i.e., at
+                          (tn, yn) = (t0, y0) or (tR, yR)
+
+     ARK_FULLRHS_END   -> called at the end of a successful step i.e, at
+                          (tcur, ycur) or the start of the subsequent step i.e.,
+                          at (tn, yn) = (tcur, ycur) from the end of the last
+                          step
+
+     ARK_FULLRHS_OTHER -> called elsewhere (e.g. for dense output)
+
+  In ForcingStep, we accumulate the RHS functions in ARK_FULLRHS_OTHER mode.
+  Generally, inner steppers will not have the correct yn when this function is
+  called and will not be able to reuse a function evaluation since their state
+  resets at the next SUNStepper_Evolve call.
+  ----------------------------------------------------------------------------*/
+static int forcingStep_FullRHS(const ARKodeMem ark_mem, const sunrealtype t,
+                               const N_Vector y, const N_Vector f,
```

Comment:
> Since const does not have much effect when applied to an `N_Vector`

It has the benefit of preventing reassignment, e.g., `y = NULL`. This is following the standard and rules for const; C's const syntax for typedef'd pointers is a bit strange, though. I'm fine removing it from the example code but think it is somewhat useful in `src`.